### PR TITLE
cancel ongoing planning if search changed, or plan instantiated

### PR DESCRIPTION
**Pull request by mmandlis**
and fix race condition

### CODE REVIEW

**Comment by mmandlis** [comment]
@sjmiles I managed to reproduce the planificator exception you were telling me about, and this should fix it too.

**Comment by mmandlis** [comment]
@shans - stupid question - if Planner is garbage collector, do i need to manually stop the speculative arc pecs, or is garbage collecting them too enough?

**Comment by shans** [comment]
So I think that because they're worklets (at least on the web) then the innerPEC will persist, as an idle eventLoop in a process waiting for messages.

Which means we probably need a way to cleanly shut them down.

**Comment by mmandlis** [comment]
@shans thanks. i added disposal for the speculative arcs. PTAL

File: runtime/test/planificator-test.js [hunk 1]
```diff
@@ -417,4 +417,13 @@ describe('Planificator', function() {
     assert.isEmpty(planificator._current.plans);
     assert.equal(4, planChanged);
   });
+  it('cancels planning', function() {
```
Comment (shaper) [inline runtime/test/planificator-test.js]: Perhaps add a test for search changing cancelling planning underway as well?

File: runtime/planner.js [hunk 2]
```diff
@@ -142,6 +149,10 @@ export class Planner {
       });
     }
   }
+  dispose() {
+    this._relevances.forEach(relevance => relevance.newArc.dispose());
```
Comment (shaper) [inline runtime/planner.js]: Add a comment on why we need to explicitly dispose (can probably paraphrase Shane's comment from the conversation).

Comment (mmandlis) [reply runtime/planner.js]: done

File: runtime/test/planificator-test.js (hunk 1 above)
Comment (mmandlis) [reply runtime/test/planificator-test.js]: done

**Comment by mmandlis** [comment]
@shans could you, please, take a look?

**Pull request merged**